ui/raidboss/emulator/ui/Tooltip.ts

```diff
@@ -1,44 +1,103 @@
+const hideEvents = [
+  'mouseleave',
+  'blur',
+] as const;
+
+const validDirections = [
+  'top',
+  'right',
+  'bottom',
+  'left',
+] as const;
+type ValidDirection = typeof validDirections[number];
+
+abstract class HTMLTemplateElement extends HTMLElement {
+  abstract get content(): DocumentFragment;
+}
+
+const showEvents = [
+  'mouseenter',
+  'focus',
+] as const;
+
+type TemplatesType = {[Property in ValidDirection]: HTMLTemplateElement};
+
+const toPx = (px: number): string => `${px}px`;
+
 export default class Tooltip {
-  constructor(target, direction, text, autoShow = true, autoHide = true) {
-    if (!Tooltip.validDirections.includes(direction))
-      throw new Error('Invalid direction for tooltip: ' + direction);
+  private offset = {
+    x: 0,
+    y: 0,
+  };
+  private target: HTMLElement;
+  private direction: ValidDirection;
+  private tooltip: HTMLElement;
+  private inner: HTMLElement;
+  private arrow: HTMLElement;
 
-    if (typeof target === 'string')
-      target = document.querySelector(target);
+  private static templates: TemplatesType;
+
+  constructor(
+      targetRef: string | HTMLElement,
+      direction: ValidDirection,
+      text: string,
+      autoShow = true,
+      autoHide = true) {
+    Tooltip.initializeTemplates();
+
+    let target: HTMLElement | null;
+
+    if (typeof targetRef === 'string')
+      target = document.querySelector<HTMLElement>(targetRef);
+    else
+      target = targetRef;
```

Comment:
FWIW, I think it'd be better for the caller to pass in the HTMLElement rather than having the constructor find it and fail.

Reply:
I was planning on cleaning that up after the typescript conversion for all of the upstream dependencies for this class. Maybe I should add a `// @TODO:` for it as a reminder?

Reply:
TODO sounds good, thanks!